corehq/apps/userreports/management/commands/prune_old_datasources.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from collections import defaultdict
+
+from alembic.autogenerate.api import compare_metadata
+from django.core.management.base import BaseCommand
+
+from corehq.apps.userreports.models import DataSourceConfiguration, StaticDataSourceConfiguration
+from corehq.apps.userreports.sql.adapter import metadata
+from corehq.apps.userreports.util import get_indicator_adapter
+from corehq.sql_db.connections import connection_manager
+
+from fluff.signals import get_migration_context, reformat_alembic_diffs
+
+
+class Command(BaseCommand):
+    help = "Call DROP TABLE on data sources that no longer exist"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--execute',
+            action='store_true',
+            default=False,
+            help='Not implemented: Actually call DROP TABLE',
+        )
+
+    def handle(self, **options):
+        data_sources = list(DataSourceConfiguration.all())
+        data_sources.extend(list(StaticDataSourceConfiguration.all()))
+
+        tables_by_engine = defaultdict(list)
+        for data_source in data_sources:
+            adapter = get_indicator_adapter(data_source)
+            if hasattr(adapter, 'engine_id'):
+                tables_by_engine[adapter.engine_id].append(adapter.get_table().name)
+
+        diffs_by_engine = defaultdict(list)
+        for engine_id, table_map in tables_by_engine.items():
```

Comment:
Non-blocking nit: You don't use `table_map`, so you could combine this and the next two `for` loops by just iterating the engines once. That could use the same connection to fetch the diffs and count the rows of each table to remove.

Reply:
This `defaultdict` is only a `set` in the new implementation. thanks for catching